File: src/c04_professional_classes/liskov/refactored/LiskovExampleRefactored.java
package c04_professional_classes.liskov.refactored;/*
In this example:

- We created a Shape interface with getArea as an abstract method.
- Rectangle and Square each have independent implementations of getArea.
- This design allows both Rectangle and Square to be used interchangeably in testArea without any unexpected behavior, preserving the Liskov Substitution Principle.
*/

interface Shape {
    int getArea();
}

class Rectangle implements Shape {
    protected int width;
    protected int height;

    public Rectangle(int w, int h) {
        this.width = w;
        this.height = h;
    }

    public void setWidth(int w) {
        this.width = w;
    }

    public void setHeight(int h) {
        this.height = h;
    }

    @Override
    public int getArea() {
        return width * height;
    }
}

class Square implements Shape {
    private int side;

    public Square(int s) {
        this.side = s;
    }

    public void setSide(int s) {
        this.side = s;
    }

    @Override
    public int getArea() {
        return side * side;
    }
}

public class LiskovExampleRefactored {
    public static void testArea(Shape shape) {
        System.out.println("Area: " + shape.getArea());
    }

    public static void main(String[] args) {
        Rectangle rect = new Rectangle(5, 10);
        testArea(rect);

        Square square = new Square(5);
        testArea(square);
    }
}

File: src/c04_professional_classes/matlab/LiskovExampleRefactored.m
function LiskovExampleRefactored()
    % In this example:
    % - We created a Shape interface (abstract base class) with getArea.
    % - RectangleRef and SquareRef each have independent implementations of getArea.
    % - This design allows both Rectangle and Square to be used interchangeably
    %   in testArea without any unexpected behavior, preserving the LSP.

    rect = RectangleRef(5, 10);
    testArea(rect);

    square = SquareRef(5);
    testArea(square);
end

function testArea(shape)
    fprintf('Area: %d\n', shape.getArea());
end

classdef IShape < handle
    methods (Abstract)
        a = getArea(obj)
    end
end

classdef RectangleRef < IShape
    properties (Access = protected)
        width
        height
    end
    methods
        function obj = RectangleRef(w, h)
            obj.width = w;
            obj.height = h;
        end
        function setWidth(obj, w)
            obj.width = w;
        end
        function setHeight(obj, h)
            obj.height = h;
        end
        function a = getArea(obj)
            a = obj.width * obj.height;
        end
    end
end

classdef SquareRef < IShape
    properties (Access = private)
        side
    end
    methods
        function obj = SquareRef(s)
            obj.side = s;
        end
        function setSide(obj, s)
            obj.side = s;
        end
        function a = getArea(obj)
            a = obj.side * obj.side;
        end
    end
end
